# Home

## MANIPULATION OF NGS DATA FOR GENOMIC AND POPULATION GENETICS ANALYSES

* Course â€“ 3rd Edition
* January 18th-22nd, 2021

<a href="https://www.transmittingscience.com/courses/genetics-and-genomics/manipulation-ngs-data-genomic-population-genetics-analyses/" target="_blank">Transmitting Science Page</a>

## Contact

courses@transmittingscience.com

## Program

###### Day 1, Monday

     * Discover the NGS world !
         Short presentation on the content of the course
         Short introduction on the 2nd generation of sequencing

     * Discovering the linux world throught jupyter world
         Presentation on using Linux for manipulationg data
         Presentation of Jupyter notebooks
         Practice in complete group
        
     * ON YOUR OWN: 
        Counting number of reads, of files
        Plot some basics stats
        Recover sequences of interest from a GFF (capture seq)
        Compare two files (BED)
        Bonus: create a mini workflow with pipe
        Solo

###### Day 2, Tuesday

    * Discover the mapping and SNP calling methodology
        Presentation: What is a mapping ? How do I call NSP ?

    * Mapping & SNP calling on one sample
        Practical in complete group

    * ON YOUR OWN: 
        Mapping and SNP calling on several samples
        Plotting density on chromosomes
        Solo or small groups


###### Day 3, Wednesday

    * Understanding the results: Ok we have a SNP file but does it mean something real ?
        Vizualiazation of an alignment file with TABLET
        Group discussion on some SNP

    * How to select the most relevant SNP, ie removing the maximum of errors ?
        Practical on SNP filtering

    * How to use these SNP for population genetics
        Practical on sNMF, a small and fast tool for population genetics
        
    * ON YOUR OWN: 
        Playing with GATK parameters for SNP filtering and impact on sNMF
        Performing PCA on SNP data
        Solo or small groups

###### Day 4, Thursday

    * Long reads 
        Introduction on the third generation, with a focus on ONT
        Some practical use cases for long reads

    * Mapping & long reads: differences with short reads ?
        Practical QC on long reads
        Practical Mapping with Minimap2 for one sample

    * Long reads and SV 
        Basic introduction on SV
        Practical SV detection with Sniffles on one sample

    * ON YOUR OWN: 
        SV detection on all samples
        Plotting SV & SNP on chromosomes
        Comparing with SNP high density regions
        Solo or small groups
    

###### Day 5, Friday

        Basic introduction on Assembly for Short & Long reads
        Difference short/long reads results & assembly comparison tools presentation
        Some more recent applications of NGS for population or SV
        Group discussion on your own project and how to use what you learned during this week
        



